paddle/framework/threadpool.h

```diff
@@ -0,0 +1,139 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <condition_variable>
+#include <cstdio>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace paddle {
+namespace framework {
+
+class ThreadPool {
+ private:
+  typedef std::function<void()> Task;
+  int num_threads_;
+  int available_;
+  bool running_;
+  bool done_;
+  std::queue<Task> tasks_;
+  std::vector<std::unique_ptr<std::thread>> threads_;
+  std::mutex mutex_;
+  std::condition_variable scheduled_;
+  std::condition_variable completed_;
+
+ public:
+  /**
+   * @brief   get a instance of threadpool, the thread number will
+   *          be specified by the first time.
+   * @param[in] num_threads   The available thread number.
+   *            If num_threads <= 0, the thread pool wil be initilized
+   *            with the number of concurrent threads supported.
+   */
+  static ThreadPool* Instance(size_t num_threads) {
+    static std::unique_ptr<ThreadPool> threadpool;
+    if (threadpool.get() == nullptr) {
```

Comment:
I believe here is a race condition (as well as GetNumThreads and GetAvailable). We can either use a mutex to protect it, or do not do lazy initialization. If it's initialized when the program starts, it will *happen before* all concurrent access, thus thread safe.

Reply:
The `num_threads_` will be initialized for the first call `GetInstance()`, and never changed, so maybe it's not a race condition, and I add a lock in function`GetAvailabel()`, thanks :)

Reply:
This is a race if when `Instance` is never called before, and `Instance` is called concurrently.

Reply:
Thanks, I think `call_once` is a good choice, and use a enhancement version in https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/platform/call_once.h